pfCard: added footer action bar and filter, and Fixed #99

dtaylor113 opened the pull request:
![image](https://cloud.githubusercontent.com/assets/12733153/10073693/c3626fa6-6298-11e5-943c-c785c7ac730b.png)

jeff-phillips-18 commented on src/card/basic/card.directive.js:
```diff
@@ -3,26 +3,140 @@
+ * @param {string} headTitle Title for the card
+ * @param {string=} subTitle Sub-Title for the card
+ * @param {boolean=} showTopBorder Show/Hide the blue top border. True shows top border, false (default) hides top border
+ * @param {boolean=} showBottomBorder Show/Hide the bottom grey line between the title and sub-title.
```
Might want to call this showTitlesSeparator or something. I expected this to mean show a border at the bottom of the card.

jeff-phillips-18 commented on src/card/basic/card.directive.js:
```diff
@@ -35,7 +149,32 @@ angular.module('patternfly.card').directive('pfCard', function () {
+      }
+
+      $scope.footerCallBackFn = function () {
+        $scope.footerCallBackResult = $scope.footer.callBackFn();
```
I know this won't get called if $scope.footer.callBackFn is not defined, but you might want to check just in case.

jeff-phillips-18 commented on src/card/basic/card.html:
```diff
@@ -1,9 +1,41 @@
+          <span class="card-pf-footer-text" ng-if="footer.text">{{footer.text}}</span>
+        </span>
+      </div>
+      <div class="card-pf-cell text-right" ng-if="filter.filters && (!filter.position || filter.position === 'footer')">
```
Should this check "filter" before "filter.filters"?

jeff-phillips-18 commented:
LGTM

waldenraines commented on src/card/basic/card.directive.js:
```diff
@@ -3,26 +3,140 @@
  *
  * @example
- <example module="patternfly.card">
+ <example module="test">
```
test?

dtaylor113 replied:
just a dummy name for the ngDoc example

waldenraines replied:
Why do all of the other examples use the module name of the directive?  Is there a reason this couldn't be used here?

dtaylor113 replied:
I thought it would be confusing since this ngDoc example isn't just using the one directive, but two.  For ngDoc examples, the module name can be anything, it's just for the ngDoc, it's like a third party app.   'test' probably isn't the best name either, I'll change it to 'demo' so it's clearer.

waldenraines replied:
:+1: sounds good, thanks.

waldenraines commented on src/card/basic/card.html:
```diff
@@ -1,9 +1,41 @@
+      <div class="card-pf-cell">
+        <h2 class="card-pf-title" ng-class="{'card-pf-filter-header': showFilterInHeader()}">{{headTitle}}</h2>
+      </div>
+      <div class="card-pf-cell text-right" style="padding-right: 0px" ng-if="showFilterInHeader()">
```
Can we get this inline style out of here?

dtaylor113 replied:
Ok, removed, wasn't needed

waldenraines commented on src/charts/trends/trends-chart.directive.js:
```diff
@@ -82,10 +82,17 @@
  </file>
  <file name="script.js">
- angular.module( 'patternfly.charts' ).controller( 'ChartCtrl', function( $scope ) {
+ angular.module( 'test', ['patternfly.charts', 'patternfly.card'] ).controller( 'ChartCtrl', function( $scope ) {
```
Why the change to the test module?

dtaylor113 replied:
For the ngDoc examples, whenever two directives are needed, they need to be listed in the module dependencies.  In this case, I wrapped a trends chart inside a card.

dtaylor113 replied:
Changed module name from 'test' to 'demo'

dtaylor113 commented:
@waldenraines, ok to merge?

waldenraines commented on src/card/basic/card.directive.js:
```diff
@@ -3,26 +3,140 @@
-      <button>Click Me</button>
-    </div>
+     <div pf-card head-title="Performance" sub-title="Last 30 Days" show-top-border="false"
+          show-titles-separator="false" style="width: 65%" footer="actionBarConfig">
```
Should we move this to `demo.css`?

dtaylor113 replied:
No, I think that would be overkill, it's not a standard, nor do I think that particular width would be reusable by other demo examples.

waldenraines commented on src/card/basic/card.directive.js:
```diff
@@ -3,26 +3,140 @@
+     </div>
+
+     <div pf-card head-title="Card Title" sub-title="Card Subtitle" show-top-border="true"
+          footer="footerConfig" filter="filterConfig" style="width: 50%">
```
Should we move this inline style to `demo.css`?

dtaylor113 replied:
Same reason as above, it's just for this ngDoc example.  Each example may have different layouts, widths, etc..

waldenraines commented:
Clicking "add new cluster" here doesn't do anything, not sure if that's desired:

![screen shot 2015-09-25 at 10 15 48 am](https://cloud.githubusercontent.com/assets/4116405/10102805/74f64d7a-636e-11e5-855c-93411db32208.png)

waldenraines commented:
I am not a fan of the way this footer is declared with a config object.  In my opinion the footer should be it's own directive and be transclude the content of the directive.  Something like this:

```html
<div pf-card head-title="Performance" sub-title="Last 30 Days" show-top-border="false" show-titles-separator="false">
  <div pf-trends-chart config="configVirtual" chart-data="dataVirtual"></div>
  <div pf-trends-chart config="configPhysical" chart-data="dataPhysical"></div>
  <div pf-trends-chart config="configMemory" chart-data="dataMemory"></div>
  <div pf-card-footer>
    <i class="some-class"></i>
    <span>Whatever I want</span>
  </div>
</div>
```

The way we're doing it now is an anti-pattern and is not flexible.  I believe the footer should be created declaratively instead of using some arbitrary configuration object.  What happens when someone wants to put custom HTML into a footer instead of just text?

dtaylor113 commented:
@waldenraines  Hi, does my request to file a separate issues for the customization make sense?  If so, can you please merge this -thanks

waldenraines commented:
> @waldenraines Hi, does my request to file a separate issues for the customization make sense? If so, can you please merge this -thanks

I never saw that request because you edited my comment instead of posting a new one.  In the future, you can highlight text and hit the "r" button to quote like I am doing above.

See https://github.com/patternfly/angular-patternfly/issues/105 for the requested issue.

waldenraines commented:
I still feel as if we could accomplish what has been designed more elegantly by using my suggestion but since we will already have to update the header similarly I am fine with you merging this.

dtaylor113 commented:
> I never saw that request because you edited my comment instead of posting a new one (and thus no email notification). In the future, you can highlight text and hit the "r" key to quote like I am doing above.

Ah..good to know, was wondering about that.  sometimes I see "Add a Note Inline" button, but other times not.   Anyway, if I have addressed all your issues, can you please merge?  -thanks

waldenraines commented:
> sometimes I see "Add a Note Inline" button, but other times not.

Yeah, GH has the "conversation view" only for comments on code and not for general comments.  Would be nice if they had it for both.
